merge_cdxj.py

```diff
@@ -0,0 +1,150 @@
+import argparse
+import json
+import logging
+import os
+import re
+import time
+from datetime import timedelta
+
+from langcodes import Language
+
+
+def get_alpha3_language_codes(lang_list):
+    """Find each language code and convert it to alpha3 using langcodes."""
+    code = []
+    for dict in lang_list:
+        new_code = ''
+        lang_code = dict['code']
+        try:
+            new_code = Language.get(lang_code).to_alpha3()
+        except LookupError as err:
+            logging.error(err)
+        # We only want to include the language if it has a 3 letter code.
+        if len(new_code) == 3:
+            code.append(new_code)
+    # The codes need to be comma separated values.
+    lang_codes = ','.join(code)
+    return lang_codes
+
+
+def get_sidecar_fields(original_obj, meta_obj):
+    """Collect the mime, charset, languages, and soft404 to add them to the original WARC dict."""
+    if meta_obj.get('Identified-Payload-Type'):
+        # Choosing python-magic over fido, due to broader choices.
+        if meta_obj['Identified-Payload-Type'].get('python-magic'):
+            mime = meta_obj['Identified-Payload-Type']['python-magic']
+        else:
+            mime = meta_obj['Identified-Payload-Type']['fido']
+        original_obj['mime-detected'] = mime
+    if meta_obj.get('Charset-Detected'):
+        charset = meta_obj['Charset-Detected']['encoding']
+        original_obj['charset'] = charset
+    if meta_obj.get('Languages-cld2'):
+        lang_array = meta_obj['Languages-cld2']['languages']
+        new_codes = get_alpha3_language_codes(lang_array)
+        if new_codes:
+            original_obj['languages'] = new_codes
+    if meta_obj.get('Soft-404-Detected'):
+        soft404 = meta_obj['Soft-404-Detected']
+        original_obj['soft-404-detected'] = soft404
+    return original_obj
+
+
+def merge_meta_fields(meta_dict, original_cdxj):
+    """Find the matching keys, merge the JSON objects, then update the line for the new CDXJ."""
+    edited_count = 0
+    non_edited_count = 0
+    list_of_original = []
+    for line in original_cdxj:
```

Comment:
At first I was converting original_cdxj into a dictionary, but I had to account for possibly having doubles (with the same surt url and timestamp) and keeping track of those in a separate dictionary... and hoping there were only doubles and not more than 2 of the same key. The way I refactored became a bit cleaner and less code. This way is only concerned with adding the sidecar fields into any matching urlkey+timestamp from the original and it doesn't care that the original may have more than one of the same urlkey+timestamp.

Hopefully that explains what I'm trying to do and there could still be a better way, but so far this works and doesn't have me creating and keeping track of several dictionaries for one file.

Reply:
ah ok, I believe I see what you're saying. With that clarification, I think this approach makes sense. I wonder if maybe a comment might be warranted here, saying that the reason for the loop is so that we add all the additional info from meta_dict to _each_ matching urlkey+timestamp line (whatever it's called) in original_cdxj, as there may be duplicates.

Reply:
Also, I don't recall seeing that scenario (adding the extra info to all of 2+ lines with the same urlkey+timestamp) covered in the tests, but it seems like it was a big consideration for the design decisions made for that function. Maybe we should have a test to show that behavior?